examples/local/vstream_client.go

```diff
@@ -82,17 +74,375 @@ func main() {
+
+// getLastVgtid retrieves the last vgtid processed by the client, so that it can resume from that position.
+func getLastVgtid(ctx context.Context) (*binlogdatapb.VGtid, error) {
+	var vgtid binlogdatapb.VGtid
+
+	// if storeLastVgtid was implemented, you would retrieve the last vgtid and unmarshal it here
+	// err := json.Unmarshal([]byte{}, &vgtid)
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	streamCustomer := true
+	if streamCustomer {
+		vgtid = binlogdatapb.VGtid{
+			ShardGtids: []*binlogdatapb.ShardGtid{{
+				Keyspace: "customer",
+				Shard:    "-80",
+				// Gtid "" is to stream from the start, "current" is to stream from the current gtid
+				// you can also specify a gtid to start with.
+				Gtid: "", // "current"  // "MySQL56/36a89abd-978f-11eb-b312-04ed332e05c2:1-265"
+			}, {
+				Keyspace: "customer",
+				Shard:    "80-",
+				Gtid:     "",
+			}},
+		}
+	} else {
+		vgtid = binlogdatapb.VGtid{
+			ShardGtids: []*binlogdatapb.ShardGtid{{
+				Keyspace: "commerce",
+				Shard:    "0",
+				Gtid:     "",
+			}},
+		}
+	}
+
+	return &vgtid, nil
+}
+
+// storeLastVgtid stores the last vgtid processed by the client, so that it can resume from that position on restart.
+// Storing a json blob in a database is just one way to do this, you could put it anywhere.
+func storeLastVgtid(ctx context.Context, vgtid *binlogdatapb.VGtid) error {
+	_, err := json.Marshal(vgtid)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Customer is the concrete type that will be built from the stream
+type Customer struct {
+	ID    int64
+	Email string
+
+	// the fields below aren't actually in the schema, but are added for illustrative purposes
+	EmailConfirmed bool
+	Details        map[string]any
+	CreatedAt      time.Time
+}
+
+// to avoid flushing too often, we will only flush if it has been at least minFlushDuration since the last flush.
+// we're relying on heartbeat events to handle max duration between flushes, in case there are no other events.
+const minFlushDuration = 5 * time.Second
+
+// maxCustomersToFlush serves two purposes:
+//  1. it limits the number of customers that are flushed at once, to avoid large transactions. If more than this number
+//     of customers are processed, they will be flushed in chunks of this size.
+//  2. if this number is exceeded before reaching the minFlushDuration, it will trigger a flush to avoid holding too much
+//     memory in the customers slice.
+const maxCustomersToFlush = 1000
+
+func readEvents(ctx context.Context, reader vtgateconn.VStreamReader, heartbeatDur time.Duration) error {
+	// customers built from the stream will be stored in this slice, until it's time to flush them to the database,
+	// at which point they will be cleared and the slice reused
+	customers := make([]*Customer, 0, maxCustomersToFlush)
+
+	// the first events will be field events, which contains schema information for any tables that are being streamed
+	var customerFields, corderFields []*querypb.Field
```

Comment:
As I'm thinking, these are per shard, and this only works if schemas are identical across all shards. That may be fine here, but you'd have to account for it in a production setting. Maybe keep a mutex guarded `map[keyspace/shard.table]*Fields`?